File: free-small-fft-in-multiple-languages/fft-test.c
/* 
 * FFT and convolution test (C)
 * 
 * Copyright (c) 2017 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/free-small-fft-in-multiple-languages
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

#include <math.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "fft.h"


// Private function prototypes
static void test_fft(int n);
static void test_convolution(int n);
static void naive_dft(const double *inreal, const double *inimag,
	double *outreal, double *outimag, bool inverse, int n);
static void naive_convolve(const double *xreal, const double *ximag,
	const double *yreal, const double *yimag,
	double *outreal, double *outimag, int n);
static double log10_rms_err(const double *xreal, const double *ximag,
	const double *yreal, const double *yimag, int n);
static double *random_reals(int n);
static void *memdup(const void *src, size_t n);

static double max_log_error = -INFINITY;


/*---- Main and test functions ----*/

int main(void) {
	srand(time(NULL));
	
	// Test power-of-2 size FFTs
	for (int i = 0; i <= 12; i++)
		test_fft(1 << i);
	
	// Test small size FFTs
	for (int i = 0; i < 30; i++)
		test_fft(i);
	
	// Test diverse size FFTs
	for (int i = 0, prev = 0; i <= 100; i++) {
		int n = (int)lround(pow(1500, i / 100.0));
		if (n > prev) {
			test_fft(n);
			prev = n;
		}
	}
	
	// Test power-of-2 size convolutions
	for (int i = 0; i <= 12; i++)
		test_convolution(1 << i);
	
	// Test diverse size convolutions
	for (int i = 0, prev = 0; i <= 100; i++) {
		int n = (int)lround(pow(1500, i / 100.0));
		if (n > prev) {
			test_convolution(n);
			prev = n;
		}
	}
	
	printf("\n");
	printf("Max log err = %.1f\n", max_log_error);
	printf("Test %s\n", max_log_error < -10 ? "passed" : "failed");
	return EXIT_SUCCESS;
}


static void test_fft(int n) {
	double *inputreal = random_reals(n);
	double *inputimag = random_reals(n);
	
	double *refoutreal = malloc(n * sizeof(double));
	double *refoutimag = malloc(n * sizeof(double));
	naive_dft(inputreal, inputimag, refoutreal, refoutimag, false, n);
	
	double *actualoutreal = memdup(inputreal, n * sizeof(double));
	double *actualoutimag = memdup(inputimag, n * sizeof(double));
	Fft_transform(actualoutreal, actualoutimag, n);
	
	printf("fftsize=%4d  logerr=%5.1f\n", n,
		log10_rms_err(refoutreal, refoutimag, actualoutreal, actualoutimag, n));
	
	free(inputreal);
	free(inputimag);
	free(refoutreal);
	free(refoutimag);
	free(actualoutreal);
	free(actualoutimag);
}


static void test_convolution(int n) {
	double *input0real = random_reals(n);
	double *input0imag = random_reals(n);
	double *input1real = random_reals(n);
	double *input1imag = random_reals(n);
	
	double *refoutreal = malloc(n * sizeof(double));
	double *refoutimag = malloc(n * sizeof(double));
	naive_convolve(input0real, input0imag, input1real, input1imag, refoutreal, refoutimag, n);
	
	double *actualoutreal = malloc(n * sizeof(double));
	double *actualoutimag = malloc(n * sizeof(double));
	Fft_convolveComplex(input0real, input0imag, input1real, input1imag, actualoutreal, actualoutimag, n);
	
	printf("convsize=%4d  logerr=%5.1f\n", n,
		log10_rms_err(refoutreal, refoutimag, actualoutreal, actualoutimag, n));
	
	free(input0real);
	free(input0imag);
	free(input1real);
	free(input1imag);
	free(refoutreal);
	free(refoutimag);
	free(actualoutreal);
	free(actualoutimag);
}


/*---- Naive reference computation functions ----*/

static void naive_dft(const double *inreal, const double *inimag,
		double *outreal, double *outimag, bool inverse, int n) {
	
	double coef = (inverse ? 2 : -2) * M_PI;
	for (int k = 0; k < n; k++) {  // For each output element
		double sumreal = 0;
		double sumimag = 0;
		for (int t = 0; t < n; t++) {  // For each input element
			double angle = coef * ((long long)t * k % n) / n;
			sumreal += inreal[t] * cos(angle) - inimag[t] * sin(angle);
			sumimag += inreal[t] * sin(angle) + inimag[t] * cos(angle);
		}
		outreal[k] = sumreal;
		outimag[k] = sumimag;
	}
}


static void naive_convolve(const double *xreal, const double *ximag,
		const double *yreal, const double *yimag,
		double *outreal, double *outimag, int n) {
	
	for (int i = 0; i < n; i++) {
		outreal[i] = 0;
		outimag[i] = 0;
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			int k = (i + j) % n;
			outreal[k] += xreal[i] * yreal[j] - ximag[i] * yimag[j];
			outimag[k] += xreal[i] * yimag[j] + ximag[i] * yreal[j];
		}
	}
}


/*---- Utility functions ----*/

static double log10_rms_err(const double *xreal, const double *ximag,
		const double *yreal, const double *yimag, int n) {
	
	double err = pow(10, -99 * 2);
	for (int i = 0; i < n; i++) {
		double real = xreal[i] - yreal[i];
		double imag = ximag[i] - yimag[i];
		err += real * real + imag * imag;
	}
	
	err /= n > 0 ? n : 1;
	err = sqrt(err);  // Now this is a root mean square (RMS) error
	err = log10(err);
	if (err > max_log_error)
		max_log_error = err;
	return err;
}


static double *random_reals(int n) {
	double *result = malloc(n * sizeof(double));
	for (int i = 0; i < n; i++)
		result[i] = (rand() / (RAND_MAX + 1.0)) * 2 - 1;
	return result;
}


static void *memdup(const void *src, size_t n) {
	void *dest = malloc(n);
	if (dest != NULL)
		memcpy(dest, src, n);
	return dest;
}

File: free-small-fft-in-multiple-languages/fft-test.rs
/* 
 * FFT and convolution test (Rust)
 * 
 * Copyright (c) 2020 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/free-small-fft-in-multiple-languages
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

extern crate rand;
use rand::distributions::IndependentSample;
mod fft;


/*---- Main and test functions ----*/

fn main() {
	let mut maxlogerr: f64 = std::f64::NEG_INFINITY;
	{
		let mut update_error = |err: f64| maxlogerr = maxlogerr.max(err);
		
		// Test power-of-2 size FFTs
		for i in 0 .. 13 {
			update_error(test_fft(1usize << i));
		}
		
		// Test small size FFTs
		for i in 0 .. 30 {
			update_error(test_fft(i));
		}
		
		// Test diverse size FFTs
		let mut prev: usize = 0;
		for i in 0i32 .. 101 {
			let n: usize = (1500.0f64).powf(f64::from(i) / 100.0).round() as usize;
			if n > prev {
				update_error(test_fft(n));
				prev = n;
			}
		}
		
		// Test power-of-2 size convolutions
		for i in 0 .. 13 {
			update_error(test_convolution(1usize << i));
		}
		
		// Test diverse size convolutions
		let mut prev: usize = 0;
		for i in 0i32 .. 101 {
			let n: usize = (1500.0f64).powf(f64::from(i) / 100.0).round() as usize;
			if n > prev {
				update_error(test_convolution(n));
				prev = n;
			}
		}
	}
	
	println!();
	println!("Max log err = {:.1}", maxlogerr);
	println!("Test {}", if maxlogerr < -10.0 { "passed" } else { "failed" });
}


fn test_fft(size: usize) -> f64 {
	let inputreal: Vec<f64> = random_reals(size);
	let inputimag: Vec<f64> = random_reals(size);
	
	let (expectreal, expectimag): (Vec<f64>,Vec<f64>) =
		naive_dft(&inputreal, &inputimag, false);
	
	let mut actualreal: Vec<f64> = inputreal.clone();
	let mut actualimag: Vec<f64> = inputimag.clone();
	fft::transform(&mut actualreal, &mut actualimag);
	let mut err: f64 = log10_rms_err(&expectreal, &expectimag, &actualreal, &actualimag);
	
	for x in actualreal.iter_mut() {
		*x /= size as f64;
	}
	for x in actualimag.iter_mut() {
		*x /= size as f64;
	}
	fft::inverse_transform(&mut actualreal, &mut actualimag);
	err = log10_rms_err(&inputreal, &inputimag, &actualreal, &actualimag).max(err);
	println!("fftsize={:4}  logerr={:5.1}", size, err);
	err
}


fn test_convolution(size: usize) -> f64 {
	let input0real: Vec<f64> = random_reals(size);
	let input0imag: Vec<f64> = random_reals(size);
	let input1real: Vec<f64> = random_reals(size);
	let input1imag: Vec<f64> = random_reals(size);
	
	let (expectreal, expectimag): (Vec<f64>,Vec<f64>) =
		naive_convolve(&input0real, &input0imag, &input1real, &input1imag);
	
	let (actualreal, actualimag) = fft::convolve_complex(input0real, input0imag, input1real, input1imag);
	
	let err: f64 = log10_rms_err(&expectreal, &expectimag, &actualreal, &actualimag);
	println!("convsize={:4}  logerr={:5.1}", size, err);
	err
}


/*---- Naive reference computation functions ----*/

fn naive_dft(inreal: &[f64], inimag: &[f64], inverse: bool) -> (Vec<f64>,Vec<f64>) {
	let n: usize = inreal.len();
	assert_eq!(inreal.len(), n);
	let mut outreal = vec![0.0f64; n];
	let mut outimag = vec![0.0f64; n];
	
	let coef: f64 = if inverse { 2.0 } else { -2.0 } * std::f64::consts::PI;
	for k in 0 .. n {  // For each output element
		let mut sumreal: f64 = 0.0;
		let mut sumimag: f64 = 0.0;
		for t in 0 .. n {  // For each input element
			let temp: u64 = (t as u64) * (k as u64) % (n as u64);  // This is more accurate than t * k
			let angle: f64 = coef * (temp as f64) / (n as f64);
			sumreal += inreal[t] * angle.cos() - inimag[t] * angle.sin();
			sumimag += inreal[t] * angle.sin() + inimag[t] * angle.cos();
		}
		outreal[k] = sumreal;
		outimag[k] = sumimag;
	}
	(outreal, outimag)
}


fn naive_convolve(
		xreal: &[f64], ximag: &[f64],
		yreal: &[f64], yimag: &[f64])
		-> (Vec<f64>,Vec<f64>) {
	
	let n: usize = xreal.len();
	assert_eq!(ximag.len(), n);
	assert_eq!(yreal.len(), n);
	assert_eq!(yimag.len(), n);
	
	let mut outreal = vec![0.0f64; n];
	let mut outimag = vec![0.0f64; n];
	for i in 0 .. n {
		for j in 0 .. n {
			let k: usize = (i + j) % n;
			outreal[k] += xreal[i] * yreal[j] - ximag[i] * yimag[j];
			outimag[k] += xreal[i] * yimag[j] + ximag[i] * yreal[j];
		}
	}
	(outreal, outimag)
}


/*---- Utility functions ----*/

fn log10_rms_err(
		xreal: &[f64], ximag: &[f64],
		yreal: &[f64], yimag: &[f64]) -> f64 {
	
	let n: usize = xreal.len();
	assert_eq!(ximag.len(), n);
	assert_eq!(yreal.len(), n);
	assert_eq!(yimag.len(), n);
	
	let mut err: f64 = (10.0f64).powf(-99.0 * 2.0);
	for i in 0 .. n {
		let real: f64 = xreal[i] - yreal[i];
		let imag: f64 = ximag[i] - yimag[i];
		err += real * real + imag * imag;
	}
	// Calculate root mean square (RMS) error
	err /= std::cmp::max(n, 1) as f64;
	err.sqrt().log10()
}


fn random_reals(size: usize) -> Vec<f64> {
	let rng = &mut rand::thread_rng();
	let uniform = rand::distributions::range::Range::new(-1.0, 1.0);
	(0 .. size).map(|_| uniform.ind_sample(rng)).collect()
}
